### PR TITLE
Exclusive use of --update --replace

### LINKED ISSUE #294: Exclusive use of --update --replace
`deploy` and `store deploy` have `--update` and  `--replace` flags, which should be strictly exclusive to ensure consistency of these flags

## Current Behaviour
Valid
```
--update=false --replace=true
--update=true --replace=false
--update=false --replace=false
```

Invalid
```
--replace=true
--update=true --replace=true
```

## New Behaviour
Valid
```
--replace=true
--update=true
--replace=false
--update=false (both false, whtat to do? I'll get back)
```

## Possible Solution
User should use one of these flags at once.
If user use `--replace`, `--update` changes its default value to false

### PR DESCRIPTION
**Pull request by itscaro**
these flags are no longer accepted at the same time even their values are different
and
when --replace is used, --update is set to false

## Motivation and Context

- [x] I have raised an issue to propose this change (closes #294)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: commands/deploy.go [hunk 1]
```diff
@@ -26,38 +27,18 @@ type DeployFlags struct {
...
-
-	faasCmd.AddCommand(deployCmd)
+	faasCmd.AddCommand(newDeployCmd())
 }
 
-// deployCmd handles deploying OpenFaaS function containers
-var deployCmd = &cobra.Command{
-	Use: `deploy -f YAML_FILE [--replace=false]
+// newDeployCmd handles deploying OpenFaaS function containers
+func newDeployCmd() *cobra.Command {
```
Comment (itscaro) [inline commands/deploy.go]: This re-factor is needed to run test

**Comment by alexellis** [comment]
I'm a bit confused with this and the original issue. What is happening exactly in the PR?

File: commands/deploy.go [hunk 2]
```diff
@@ -341,3 +376,15 @@ func deriveFprocess(function stack.Function) (string, error) {
 func languageExistsNotDockerfile(language string) bool {
 	return len(language) > 0 && strings.ToLower(language) != "dockerfile"
 }
+
+// ensureUpdateReplaceFlags ensures the consistency of 'replace' and 'update' flags
+func ensureUpdateReplaceFlags(cmd *cobra.Command, deployFlags *DeployFlags) error {
```
Comment (itscaro) [inline commands/deploy.go]: Validates --update & --replace

File: commands/store.go [hunk 3]
```diff
@@ -99,7 +99,16 @@ var storeDeployCmd = &cobra.Command{
 							faas-cli store deploy figlet
 									--gateway=http://remote-site.com:8080 --lang=python
 									--env=MYVAR=myval`,
-	RunE: runStoreDeploy,
+	PreRunE: preRunStoreDeploy,
+	RunE:    runStoreDeploy,
+}
+
+// preRunStoreDeploy validates flags & args
+func preRunStoreDeploy(cmd *cobra.Command, args []string) error {
```
Comment (itscaro) [inline commands/store.go]: Move validation to PreRun to make is more clear, IMO

File: commands/deploy.go [hunk 4]
```diff
@@ -207,7 +228,21 @@ func RunDeploy(
 			return fmt.Errorf("error parsing labels: %v", labelErr)
 		}
 		functionResourceRequest1 := proxy.FunctionResourceRequest{}
-		proxy.DeployFunction(fprocess, gateway, functionName, image, language, deployFlags.replace, envvars, network, deployFlags.constraints, deployFlags.update, deployFlags.secrets, labelMap, functionResourceRequest1)
+		proxy.DeployFunction(
```
Comment (itscaro) [inline commands/deploy.go]: For readability

File: commands/deploy.go [hunk 5]
```diff
@@ -74,43 +55,70 @@ var deployCmd = &cobra.Command{
...
+	return nil
 }
 
 func runDeploy(cmd *cobra.Command, args []string) error {
-	return RunDeploy(args, image, fprocess, functionName, deployFlags)
+	return deploy(args, image, fprocess, functionName, deployFlags)
 }
 
-func RunDeploy(
+func deploy(
```
Comment (itscaro) [inline commands/deploy.go]: Maybe a bad idea to make this function local

**Comment by itscaro** [comment]
In the PR the most important change is the addition of `ensureUpdateReplaceFlags` and it's called in `PreRunE` of the command to make it clear that there are validation operations.

I needed to do refactoring, so that I can do tests. There is no way to reset Flags once set, it would be more flexible to be able to instantiate the command when needed.

**Comment by alexellis** [comment]
This PR is over a month old and in need of a rebase so I'm closing it. Feel free to propose a new PR or to rebase.

**Pull request closed without merging**